lib/location/method_gnss.c

```diff
@@ -370,93 +391,126 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
+	 * request.
+	 */
+	if (agps_req_timestamp != 0) {
+		time_since_agps_req = k_uptime_get() - agps_req_timestamp;
+		if (time_since_agps_req < (AGPS_REQUEST_INTERVAL * MSEC_PER_SEC)) {
+			LOG_DBG("Skipping A-GPS request, time since last request: %d s",
+				time_since_agps_req / 1000);
+			return false;
+		}
+	}
+
+	return true;
 }
 #endif
 
 #if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
-static void method_gnss_request_assistance(void)
+static void method_gnss_assistance_request(void)
 {
-	int err = nrf_modem_gnss_read(&agps_request,
-				      sizeof(agps_request),
-				      NRF_MODEM_GNSS_DATA_AGPS_REQ);
-
-	if (err) {
-		LOG_WRN("Reading A-GPS req data from GNSS failed, error: %d", err);
-		return;
-	}
-
 	if (IS_ENABLED(CONFIG_NRF_CLOUD_PGPS)) {
-		/* ephemerides come from P-GPS; almanacs not desired in this configuration */
+		/* Ephemerides come from P-GPS; almanacs not needed in this configuration. */
 		pgps_agps_request.sv_mask_ephe = agps_request.sv_mask_ephe;
+		pgps_agps_request.data_flags = agps_request.data_flags;
 		agps_request.sv_mask_ephe = 0;
 		agps_request.sv_mask_alm = 0;
-		LOG_DBG("P-GPS request from modem (ephe: 0x%08x)", pgps_agps_request.sv_mask_ephe);
+
+		if (IS_ENABLED(CONFIG_NRF_CLOUD_AGPS)) {
+			/* Time and location come from A-GPS. */
+			pgps_agps_request.data_flags = 0;
+		}
+
+		LOG_DBG("P-GPS request sv_mask_ephe: 0x%08x, data_flags 0x%02x",
+			pgps_agps_request.sv_mask_ephe, pgps_agps_request.data_flags);
 	}
 
-	LOG_DBG("A-GPS request from modem (ephe: 0x%08x alm: 0x%08x flags: 0x%02x)",
-		agps_request.sv_mask_ephe,
-		agps_request.sv_mask_alm,
-		agps_request.data_flags);
 #if defined(CONFIG_NRF_CLOUD_AGPS)
+	if (agps_request.data_flags != 0) {
+		/* With A-GPS, if any of the flags in the data_flags field is set, it's
+		 * advantageous to request everything at the same time because the amount of
+		 * data is so small.
+		 */
+		agps_request.data_flags =
+			NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST |
+			NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST |
+			NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST |
+			NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST |
+			NRF_MODEM_GNSS_AGPS_POSITION_REQUEST;
+	}
+
+	LOG_DBG("A-GPS request sv_mask_ephe: 0x%08x, sv_mask_alm: 0x%08x, data_flags: 0x%02x)",
+		agps_request.sv_mask_ephe, agps_request.sv_mask_alm, agps_request.data_flags);
+
 	/* Check the request. If no A-GPS data types except ephemeris or almanac are requested,
-	 * jump to P-GPS (if enabled)
+	 * jump to P-GPS (if enabled).
 	 */
 	if (method_gnss_agps_required(&agps_request)) {
+		enum lte_lc_nw_reg_status reg_status = LTE_LC_NW_REG_NOT_REGISTERED;
+
+		lte_lc_nw_reg_status_get(&reg_status);
+		if (reg_status == LTE_LC_NW_REG_REGISTERED_HOME ||
+		    reg_status == LTE_LC_NW_REG_REGISTERED_ROAMING) {
+			/* Only store the timestamp when LTE is connected, otherwise it's not
+			 * very likely that the A-GPS request succeeds.
+			 */
+			agps_req_timestamp = k_uptime_get();
+		}
```

Comment:
else: should we set agps_req_timestamp to zero?

Reply:
I don't think we ever want to zero the timestamp, it can always keep the timestamp of the last A-GPS request. If execution gets to this point, the `agps_req_timestamp` is either still zero or more than one hour has passed since the previous A-GPS request. In neither case the timestamp restricts sending of an A-GPS request.